kwanj-k/storemanager-v2
# #161404774 Add all product related endpoints

### What does this PR do?
Adds all the product related endpoints

#### Description of Task to be Completed?
 * The PR adds the following endpoints

| EndPoint                                                   | Functionality                                       |
| ------------------------------------------------------|:-------------------------------------------------:|                                                                 
| POST     /products/                                   | Add a product                                     | 
|GET       /products/                                    | Get all products                                   | 
| PUT      /products/Id/                                | Update the information of a product    |
| POST      /products/Id/                             |    Add a product to cart                         |
| DELETE   /products/Id/                            | Remove a product                                |
| GET      /cart/                                            | Get my cart                                           |
| POST      /cart/                                          | Sell an entire cart                                 |
| PUT      /cart/Id/                                         | Update a product in a cart                    |
| DELETE   /cart/id                                      | Delete a product from a cart                  |
| DELETE   /cart/                                         | Delete an entire cart                               |
| GET      /sales/                                          | Get all sales                                            |
| POST     /categories/                                |  Create a category                                   | 
|GET       /categories/                                 | Get all categories                                     | 
| PUT      /categories/Id/                             | Update a category                                    |
| DELETE   /categories/Id/                         | Remove a category                                   |
| POST    /categories/Id//products/Id/        |  Give a product a given category               |

### How should this be manually tested?
* After cloning the repo, git pull the branch ft-products-#161404774
* Install a virtual enviroment and activate it [Refer here](https://docs.python-guide.org/dev/virtualenvs/)
* Pip install the the requirements.txt
* Run the following:
  ``` export FLASK_APP='run.py' ```
  ``` export APP_SETTINGS='development' ```
   ``` pytest --cov-report term-missing --cov=app ```

Also, you need to create a development and test database and connect to it.[Refer here](https://www.digitalocean.com/community/tutorials/how-to-install-and-use-postgresql-on-ubuntu-18-04#creating-a-new-role)

Then export the database URL in the environment:
```
export Dev_URL="dbname=dbname user=username password=password"
export Test_URL="dbname=dbname user=username password=password"
export SECRET="iamasecrettheunsadtheunsulliedtheundead"

```
We need the secret key for generating the tokens

Then Run the tests:
```
pytest --cov-report term-missing --cov=app
```
* All the tests should pass:
![screenshot from 2018-10-25 20-23-04](https://user-images.githubusercontent.com/33450849/47518503-d0392c00-d893-11e8-8b54-cae7387c136f.png)



### What are the relevant Pivotal Tracker stories?
#161404774
